rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,314 @@
+    /// likely indicates a bug inside the Rerun SDK.
+    template <typename TComponent>
     class ComponentBatch {
       public:
-        const ComponentType* data;
-        size_t num_instances;
+        using TComponentType = TComponent;
 
-      public:
-        /// Construct from a single component.
+        /// Type of an adapter given input types Ts.
+        template <typename T>
+        using TAdapter =
+            ComponentBatchAdapter<TComponent, std::remove_cv_t<std::remove_reference_t<T>>>;
+
+        /// Creates a new empty component batch.
         ///
-        /// *Attention*: As with all other constructors, this does *not* take ownership of the data,
-        /// you need to ensure that the data outlives the component list.
-        ComponentBatch(const ComponentType& one_and_only) : data(&one_and_only), num_instances(1) {}
+        /// Note that logging an empty component batch is different from logging no component
+        /// batch: When you log an empty component batch at an entity that already has some
+        /// components of the same type, it will clear out all components of that type.
+        ComponentBatch() : ownership(BatchOwnership::Borrowed) {
+            storage.borrowed.data = nullptr;
+            storage.borrowed.num_instances = 0;
+        }
 
-        /// Construct from a raw pointer and size.
-        ComponentBatch(const ComponentType* _data, size_t _num_instances)
-            : data(_data), num_instances(_num_instances) {}
+        /// Construct using a `ComponentBatchAdapter`.
+        template <typename T>
+        ComponentBatch(T&& input) : ComponentBatch(TAdapter<T>()(std::forward<T>(input))) {}
 
-        /// Construct from an std::vector.
+        /// Construct from a temporary list of components.
+        ///
+        /// Persists the list into an internal std::vector.
+        /// If you want to avoid an allocation, you have to manually keep the data on the stack
+        /// (e.g. as std::array) and construct the batch from this instead.
         ///
-        /// *Attention*: As with all other constructors, this does *not* take ownership of the data,
-        /// you need to ensure that the data outlives the component list.
-        /// In particular, manipulating the passed vector after constructing the component list,
-        /// will invalidate it, similar to iterator invalidation.
-        ComponentBatch(const std::vector<ComponentType>& _data)
-            : data(_data.data()), num_instances(_data.size()) {}
-
-        /// Construct from an std::array.
+        /// This is not done as ComponentBatchAdapter since it tends to cause deduction issues.
+        ComponentBatch(std::initializer_list<TComponent> data)
+            : ownership(BatchOwnership::VectorOwned) {
+            // Don't assign, since the vector is in an undefined state and assigning may
+            // attempt to free data.
+            new (&storage.vector_owned) std::vector<TComponent>(data);
```

Comment:
this reminds me: we should really start running some runtime sanitizers on our C++ tests (ASan etc)

Reply:
https://github.com/rerun-io/rerun/issues/3797